src/ILLink.RoslynAnalyzer/COMAnalyzer.cs

```diff
@@ -0,0 +1,124 @@
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Runtime.InteropServices;
+using ILLink.Shared;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace ILLink.RoslynAnalyzer
+{
+	[DiagnosticAnalyzer (LanguageNames.CSharp)]
+	public sealed class COMAnalyzer : DiagnosticAnalyzer
+	{
+		private const string StructLayoutAttribute = nameof (StructLayoutAttribute);
+		private const string DllImportAttribute = nameof (DllImportAttribute);
+		private const string MarshalAsAttribute = nameof (MarshalAsAttribute);
+
+		static readonly DiagnosticDescriptor s_correctnessOfCOMCannotBeGuaranteed = DiagnosticDescriptors.GetDiagnosticDescriptor (DiagnosticId.CorrectnessOfCOMCannotBeGuaranteed,
+			helpLinkUri: "https://docs.microsoft.com/en-us/dotnet/core/deploying/trim-warnings/il2050");
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create (s_correctnessOfCOMCannotBeGuaranteed);
+
+		public override void Initialize (AnalysisContext context)
+		{
+			context.EnableConcurrentExecution ();
+			context.ConfigureGeneratedCodeAnalysis (GeneratedCodeAnalysisFlags.ReportDiagnostics);
+			context.RegisterCompilationStartAction (context => {
+				var compilation = context.Compilation;
+				if (!context.Options.IsMSBuildPropertyValueTrue (MSBuildPropertyOptionNames.EnableTrimAnalyzer, compilation))
+					return;
+
+				context.RegisterOperationAction (operationContext => {
+					var invocationOperation = (IInvocationOperation) operationContext.Operation;
+					var targetMethod = invocationOperation.TargetMethod;
+					if (!targetMethod.HasAttribute (DllImportAttribute))
+						return;
+
+					foreach (var parameter in targetMethod.Parameters) {
+						if (IsComInterop (parameter)) {
+							operationContext.ReportDiagnostic (Diagnostic.Create (s_correctnessOfCOMCannotBeGuaranteed,
+								operationContext.Operation.Syntax.GetLocation (), targetMethod.GetDisplayName ()));
+						}
+					}
+
+					if (IsComInterop (targetMethod.ReturnType)) {
+						operationContext.ReportDiagnostic (Diagnostic.Create (s_correctnessOfCOMCannotBeGuaranteed,
+								operationContext.Operation.Syntax.GetLocation (), targetMethod.GetDisplayName ()));
+					}
+				}, OperationKind.Invocation);
+			});
+
+			static bool IsComInterop (ISymbol symbol)
+			{
+				if (symbol.TryGetAttribute (MarshalAsAttribute, out var marshalAsAttribute) &&
+					marshalAsAttribute.ConstructorArguments.Length >= 1 && marshalAsAttribute.ConstructorArguments[0] is TypedConstant typedConstant &&
+					typedConstant.Type != null && typedConstant.Type.IsUnmanagedType) {
+					var unmanagedType = typedConstant.Value;
+					switch (unmanagedType) {
+					case (int) UnmanagedType.IUnknown:
+					case (int) UnmanagedType.IDispatch:
+					case (int) UnmanagedType.Interface:
+						return true;
+
+					default:
+						break;
```

Comment:
The linker's logic has an explicit `nativeType == NativeType.None` check - do we need something like that? (Is there a case where the marshal info can change any of the below `true` returns to be not COM related?)

Reply:
I added a check to see if the typed constant value exists in `UnmanagedType` (which mimics the check of `nativeType == NativeType.None` in Cecil).